### PR TITLE
JP- 3778 Allow the RSCD correction to work on segmented TSO data

### LINKED ISSUE #8878: RSCD step for MIRI data in the Detector 1 Pipeline is not properly handeling segmented files
_Issue [JP-3778](https://jira.stsci.edu/browse/JP-3778) was created on JIRA by [Jeroen Bouwman](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bouwman):_

The RSCD step for MIRI data in the detector 1 pipeline flags the first few groups of the detector ramps as DO_NOT_USE. Only for the first integration not flagging should be done as the first integration has no RSCD effect. However, in case of segmented files, the RSCD step treats the first integration in each segmented file as if it was the real first integration. This means that every first integration of a segment does not get flagged causing a difference in the ramp fitting and thus a difference in the photometric signal.  The RSCD step should follow the same procedure as is implemented in the dark_current step of the detector 1 pipeline for MIRI data and check if the first integration in a segmented file is the true fist integration or not by checking the  INTSTART and INTEND keywords in the header.

### PR DESCRIPTION
**Pull request by jemorrison**
Resolves [JP-3778](https://jira.stsci.edu/browse/JP-3778)

Closes #8878

This PR addresses allows the RSCD step to  work on segmented data. In segmented data the first integration in the data may not be the first integration in the full exposure. If this is the case, we want to flag initial groups in the integration as DO_NOT_USE. 

## Tasks
- [x] **request a review from someone specific**, to avoid making the maintainers review every PR
- [x] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [x] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions]
  Results: (all tests passed) https://github.com/spacetelescope/RegressionTests/actions/runs/12034294166
  - [ ] (https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
  
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.fits_generator.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

### CODE REVIEW

**Review by karllark — APPROVED** [review]
Looks great.  One minor comment.

File: jwst/rscd/rscd_sub.py [hunk 1]
```diff
@@ -96,10 +106,17 @@ def correction_skip_groups(output, group_skip):
...
+    # for integrations 1 and higher. Currently no correction is applied to first integration
+    if sci_int_start == 1:
+        output.groupdq[1:, 0:group_skip, :, :] = \
+            np.bitwise_or(output.groupdq[1:, 0:group_skip, :, :], dqflags.group['DO_NOT_USE'])
+        log.debug(f"RSCD Sub: adding DO_NOT_USE to GROUPDQ for the first {group_skip} groups")
+
+    else: # we have segmented data 
+        output.groupdq[:, 0:group_skip, :, :] = \
+            np.bitwise_or(output.groupdq[:, 0:group_skip, :, :], dqflags.group['DO_NOT_USE'])
+        log.debug(f"RSCD Sub: adding DO_NOT_USE to GROUPDQ for the first {group_skip} groups")
```
[the remark this answers is not in the document]
Comment (jemorrison) [reply jwst/rscd/rscd_sub.py:119]: good suggestion.
Done

File: jwst/rscd/rscd_sub.py [hunk 2]
```diff
@@ -80,12 +80,22 @@ def correction_skip_groups(output, group_skip):
...
+
+    if sci_int_start is None:
+        sci_int_start = 1
+        
 
     log.debug("RSCD correction using: nints=%d, ngroups=%d" %
               (sci_nints, sci_ngroups))
+    log.debug("The first integration in the data is integration: %d" %
+              (sci_int_start))
+    log.info("Number of groups to skip for integrations 2 and higher %d ", group_skip)
```
Comment (melanieclarke) [inline jwst/rscd/rscd_sub.py:98]:
```suggestion
    log.info("Number of groups to skip for integrations 2 and higher: %d " % group_skip)
```

**Review by melanieclarke — COMMENTED** [review]
Looks good, and appears to be working as expected.

@jemorrison - I know you were thinking about adding a regression test for this case, but I don't see one here.  Did you decide not to, or did it not get pushed?

**Comment by jemorrison** [comment]
@melanieclarke Somehow the regtest got left out of PR. I push the new test up

**Comment by jemorrison** [comment]
@melanieclarke   This is ready for final review

**Comment by melanieclarke** [comment]
Looks good now, and seems to be working in local tests.  Did you re-run the regression tests?

**Comment by melanieclarke** [comment]
Never mind, I see from the link above that you did re-run them yesterday, and results were clean.  I'll go ahead and approve and get this merged.  Thanks for working it through!

**Review by melanieclarke — APPROVED** [review]

**Comment by jemorrison** [comment]
@melanieclarke  Yes the link to the regression tests (above) is after I okifyed the tests yesterday.

**Pull request merged**